File: app/src/main/java/su/sres/securesms/stickers/StickerKeyboardRepository.java
package su.sres.securesms.stickers;

import android.database.Cursor;
import androidx.annotation.NonNull;

import su.sres.securesms.database.StickerDatabase;
import su.sres.securesms.database.StickerDatabase.StickerPackRecordReader;
import su.sres.securesms.database.StickerDatabase.StickerRecordReader;
import su.sres.securesms.database.model.StickerPackRecord;
import su.sres.securesms.database.model.StickerRecord;
import su.sres.core.util.concurrent.SignalExecutors;

import java.util.ArrayList;
import java.util.List;

final class StickerKeyboardRepository {

    private static final int RECENT_LIMIT = 24;

    private final StickerDatabase stickerDatabase;

    StickerKeyboardRepository(@NonNull StickerDatabase stickerDatabase) {
        this.stickerDatabase = stickerDatabase;
    }

    void getPackList(@NonNull Callback<PackListResult> callback) {
        SignalExecutors.BOUNDED.execute(() -> {
            List<StickerPackRecord> packs = new ArrayList<>();

            try (StickerPackRecordReader reader = new StickerPackRecordReader(stickerDatabase.getInstalledStickerPacks())) {
                StickerPackRecord pack;
                while ((pack = reader.getNext()) != null) {
                    packs.add(pack);
                }
            }

            boolean hasRecents;

            try (Cursor recentsCursor = stickerDatabase.getRecentlyUsedStickers(1)) {
                hasRecents = recentsCursor != null && recentsCursor.moveToFirst();
            }

            callback.onComplete(new PackListResult(packs, hasRecents));
        });
    }

    void getStickersForPack(@NonNull String packId, @NonNull Callback<List<StickerRecord>> callback) {
        SignalExecutors.BOUNDED.execute(() -> {
            List<StickerRecord> stickers = new ArrayList<>();

            try (StickerRecordReader reader = new StickerRecordReader(stickerDatabase.getStickersForPack(packId))) {
                StickerRecord sticker;
                while ((sticker = reader.getNext()) != null) {
                    stickers.add(sticker);
                }
            }

            callback.onComplete(stickers);
        });
    }

    void getRecentStickers(@NonNull Callback<List<StickerRecord>> callback) {
        SignalExecutors.BOUNDED.execute(() -> {
            List<StickerRecord> stickers = new ArrayList<>();

            try (StickerRecordReader reader = new StickerRecordReader(stickerDatabase.getRecentlyUsedStickers(RECENT_LIMIT))) {
                StickerRecord sticker;
                while ((sticker = reader.getNext()) != null) {
                    stickers.add(sticker);
                }
            }

            callback.onComplete(stickers);
        });
    }

    static class PackListResult {

        private final List<StickerPackRecord> packs;
        private final boolean                 hasRecents;

        PackListResult(List<StickerPackRecord> packs, boolean hasRecents) {
            this.packs      = packs;
            this.hasRecents = hasRecents;
        }

        List<StickerPackRecord> getPacks() {
            return packs;
        }

        boolean hasRecents() {
            return hasRecents;
        }
    }

    interface Callback<T> {
        void onComplete(T result);
    }
}

File: app/src/main/java/su/sres/securesms/keyboard/sticker/StickerKeyboardRepository.kt
package su.sres.securesms.keyboard.sticker

import android.net.Uri
import su.sres.core.util.concurrent.SignalExecutors
import su.sres.securesms.R
import su.sres.securesms.database.StickerDatabase
import su.sres.securesms.database.StickerDatabase.StickerPackRecordReader
import su.sres.securesms.database.StickerDatabase.StickerRecordReader
import su.sres.securesms.database.model.StickerPackRecord
import su.sres.securesms.database.model.StickerRecord
import java.util.function.Consumer

private const val RECENT_LIMIT = 24
private const val RECENT_PACK_ID = "RECENT"

class StickerKeyboardRepository(private val stickerDatabase: StickerDatabase) {
  fun getStickerPacks(consumer: Consumer<List<KeyboardStickerPack>>) {
    SignalExecutors.BOUNDED.execute {
      val packs: MutableList<KeyboardStickerPack> = mutableListOf()

      StickerPackRecordReader(stickerDatabase.installedStickerPacks).use { reader ->
        var pack: StickerPackRecord? = reader.next
        while (pack != null) {
          packs += KeyboardStickerPack(packId = pack.packId, title = pack.title.orNull(), coverUri = pack.cover.uri)
          pack = reader.next
        }
      }

      val fullPacks: MutableList<KeyboardStickerPack> = packs.map { p ->
        val stickers: MutableList<StickerRecord> = mutableListOf()

        StickerRecordReader(stickerDatabase.getStickersForPack(p.packId)).use { reader ->
          var sticker: StickerRecord? = reader.next
          while (sticker != null) {
            stickers.add(sticker)
            sticker = reader.next
          }
        }

        p.copy(stickers = stickers)
      }.toMutableList()

      val recentStickerPack: KeyboardStickerPack = getRecentStickerPack()
      if (recentStickerPack.stickers.isNotEmpty()) {
        fullPacks.add(0, recentStickerPack)
      }
      consumer.accept(fullPacks)
    }
  }

  private fun getRecentStickerPack(): KeyboardStickerPack {
    val recentStickers: MutableList<StickerRecord> = mutableListOf()

    StickerRecordReader(stickerDatabase.getRecentlyUsedStickers(RECENT_LIMIT)).use { reader ->
      var recentSticker: StickerRecord? = reader.next
      while (recentSticker != null) {
        recentStickers.add(recentSticker)
        recentSticker = reader.next
      }
    }

    return KeyboardStickerPack(
      packId = RECENT_PACK_ID,
      title = null,
      titleResource = R.string.StickerKeyboard__recently_used,
      coverUri = null,
      coverResource = R.drawable.ic_recent_20,
      stickers = recentStickers
    )
  }

  data class KeyboardStickerPack(
    val packId: String,
    val title: String?,
    val titleResource: Int? = 0,
    val coverUri: Uri?,
    val coverResource: Int? = null,
    val stickers: List<StickerRecord> = emptyList()
  )
}